// angular-confirmation-popover/esm2015/lib/confirmation-popover-options.provider.js
import { Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ConfirmationPopoverOptions {
    constructor() {
        this.confirmText = 'Confirm';
        this.cancelText = 'Cancel';
        this.confirmButtonType = 'success';
        this.cancelButtonType = 'outline-secondary';
        this.placement = 'top';
        this.hideConfirmButton = false;
        this.hideCancelButton = false;
        this.popoverClass = '';
        this.appendToBody = false;
        this.reverseButtonOrder = false;
        this.closeOnOutsideClick = true;
    }
}
ConfirmationPopoverOptions.ɵfac = function ConfirmationPopoverOptions_Factory(t) { return new (t || ConfirmationPopoverOptions)(); };
ConfirmationPopoverOptions.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ConfirmationPopoverOptions, factory: ConfirmationPopoverOptions.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConfirmationPopoverOptions, [{
        type: Injectable
    }], function () { return []; }, null); })();
