// angular-confirmation-popover/esm2015/lib/focus.directive.js
import { Directive, ElementRef, Input, } from '@angular/core';
/**
 * A helper directive to focus buttons. You will only need this if using a custom template
 */
import * as ɵngcc0 from '@angular/core';
export class FocusDirective {
    constructor(elm) {
        this.elm = elm;
    }
    ngOnChanges(changes) {
        if (changes.mwlFocus && this.mwlFocus === true) {
            this.elm.nativeElement.focus();
        }
    }
}
FocusDirective.ɵfac = function FocusDirective_Factory(t) { return new (t || FocusDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
FocusDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: FocusDirective, selectors: [["", "mwlFocus", ""]], inputs: { mwlFocus: "mwlFocus" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
FocusDirective.ctorParameters = () => [
    { type: ElementRef }
];
FocusDirective.propDecorators = {
    mwlFocus: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FocusDirective, [{
        type: Directive,
        args: [{
                selector: '[mwlFocus]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { mwlFocus: [{
            type: Input
        }] }); })();
